jinahub/rankers/LightGBMRanker/lightgbm_ranker.py

```diff
@@ -0,0 +1,209 @@
+            categorical_feature=self.categorical_match_features,
+            free_raw_data=False,
+        )
+        if self.query_features_before:
+            return query_dataset.construct().add_features_from(
+                match_dataset.construct()
+            )
+        else:
+            return match_dataset.construct().add_features_from(
+                query_dataset.construct()
+            )
+
+    @requests(on='/train')
+    def train(self, docs: DocumentArray, **kwargs):
+        """The :meth:`train` endpoint allows user to train the lightgbm ranker
+        in an incremental manner. The features will be extracted from the `attr`:`tags`,
+        including all the :attr:`query_features` and :attr:`match_features`. The label/groundtruth of the
+        training data will be the :attr:`label` field.
+
+        :param docs: :class:`DocumentArray` passed by the user or previous executor.
+        :param kwargs: Additional key value arguments.
+        """
+        train_set = self._get_features_dataset(docs)
+        categorical_feature = []
+        if self.categorical_query_features:
+            categorical_feature += self.categorical_query_features
+        if self.categorical_match_features:
+            categorical_feature += self.categorical_match_features
+        if not categorical_feature:
+            categorical_feature = 'auto'
+        self.booster = lightgbm.train(
+            train_set=train_set,
+            init_model=self.booster,
+            params=self.params,
+            keep_training_booster=True,
+            categorical_feature=categorical_feature,
+        )
+
+    @requests(on='/search')
+    def rank(self, docs: DocumentArray, **kwargs):
+        """The :meth:`rank` endpoint allows user to assign a score to their docs given by pre-trained
+          :class:`LightGBMRanker`. Once load, the :class:`LightGBMRanker` will load the pre-trained model
+          and make predictions on the documents. The predictions are made based on extracted dataset from
+          query and matches. The :attr:`query_features` will be extracted from query document :attr:`tags`
+          and `match_features` will be extracted from corresponded matches documents tags w.r.t the query document.
+
+        :param docs: :class:`DocumentArray` passed by the user or previous executor.
+        :param kwargs: Additional key value arguments.
+        """
+        dataset = self._get_features_dataset(docs)
+        if self.booster:
+            predictions = self.booster.predict(dataset.get_data())
+            matches = docs.traverse_flat(traversal_paths=['m'])
+            for prediction, match in zip(predictions, matches):
+                match.scores[self.label] = prediction
+            for doc in docs:
+                doc.matches.sort(key=lambda x: x.scores[self.label].value, reverse=True)
+        else:
+            self.logger.warning('Model not trained, will return initial docs.')
```

Comment:
Test this warning using `caplog` fixture (if it's not too much effort)

Reply:
i don't see it's necessary..